### PR TITLE
Add Sofa recipe

### PR DESCRIPTION
**Pull request by olivier-roussel**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by olivier-roussel** [comment]
@conda-forge/help-c-cpp @conda-forge/help-python-c The recipe is ready for review ! 
Sorry for the big commits number, it was a long work. I tried to squash a bit. If necessary, next step would be to squash all in one... (not sure about the merge and squash policy for the PR in conda-forge, so feel free to let me know if extra squashing is needed).

**Comment by carterbox** [comment]
> Please note that macOS build is temporarily disabled as the SOFA CI used to generate the current release (v23.06) artifacts, including source code archives, is currently in down and in maintenance for macOS.

Why does upstream have different source releases for different platforms? Or it's just that MacOS isn't tested for the latests release, so you don't want to publish it yet?

**Comment by carterbox** [comment]
Are the Python/C++ portions of these plugin packages separable? We typically don't like to force the python toolchain on C++ developers/package consumers if it's not necessary. I'll check back a few hours for the finished builds.

**Review by carterbox — CHANGES_REQUESTED** [review]
We do not allow revending or static-linking or shipping of static libs. SOFA is clearly doing all three of these.

- [ ] REQUIRED: All the header-only external projects need to have their licenses added to the license section of this recipe.  https://github.com/sofa-framework/sofa/tree/master/extlibs
- [ ] REQUIRED: Non-header-only external projects including (but maybe not limited to) TinyXML, CXXOpts, Metis, should use separate conda packages with dynamic links. We already have many of these on conda-forge.
- [ ] REQUIRED: libcsparse is not licensed for redistribution? The [license file for that package ](https://github.com/sofa-framework/sofa/blob/master/Sofa/Component/LinearSolver/Direct/extlibs/csparse/COPYING.txt) is not a license. It is a copyright statement. We cannot distribute code from this project in any manner until we have a valid license.
- [ ] Consider using [glob file lists](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#specifying-files-to-include-in-output) to split your package into separate outputs i.e. sofa, sofa-devel, libsofa so that end-users don't get headers, binaries, or /share that they don't need.

**Comment by olivier-roussel** [comment]
Thanks @carterbox for your review.

> Why does upstream have different source releases for different platforms? Or it's just that MacOS isn't tested for the latests release, so you don't want to publish it yet?
I might have been unclear on this point. SOFA does not have different source releases for each platform, but if my memory is correct about this, the CI of SOFA plugins will download SOFA binaries as SOFA (core) is a dependency of SOFA plugins (and some plugins are themselves a dependency of others).
MacOS is normally supported, but the support was temporarily disabled simply because the CI servers were moved and if I understood correctly there was some problems getting them back to normal (@hugtalbot correct me if I'm wrong). This took a bit of time, macOS CI is being currently restored and I hope being able to provide macOS support for the conda package soon. I was thinking it as non-blocking point for a first release on conda forge.

**Comment by olivier-roussel** [comment]
> Are the Python/C++ portions of these plugin packages separable? We typically don't like to force the python toolchain on C++ developers/package consumers if it's not necessary. I'll check back a few hours for the finished builds.

That makes sense. I guess it should be separable and joins your point with multiple outputs. I'll have a deeper look and work on it.

**Comment by olivier-roussel** [comment]
> We do not allow revending or static-linking or shipping of static libs. SOFA is clearly doing all three of these.

I totally missed the licenses embedded libraries points for some reason... I thought I didn't tick the checkbox of the checklist yet. Anyway, thanks a lot for spotting this.

>    * [ ]  REQUIRED: All the header-only external projects need to have their licenses added to the license section of this recipe.  https://github.com/sofa-framework/sofa/tree/master/extlibs

Ok, should be easy to fix.

>    * [ ]  REQUIRED: Non-header-only external projects including (but maybe not limited to) TinyXML, CXXOpts, Metis, should use separate conda packages with dynamic links. We already have many of these on conda-forge.

I will check with the SOFA developers especially how we will manage that. Might take a bit of time.

>    * [ ]  REQUIRED: libcsparse is not licensed for redistribution? The [license file for that package ](https://github.com/sofa-framework/sofa/blob/master/Sofa/Component/LinearSolver/Direct/extlibs/csparse/COPYING.txt) is not a license. It is a copyright statement. We cannot distribute code from this project in any manner until we have a valid license.

Indeed. Seems that CSparse is part of SuiteSparse framework ( https://github.com/DrTimothyAldenDavis/SuiteSparse/tree/dev/CSparse ), where the license appears clearly (LGPL 2.1). I'll update this to add the correct license.

>    * [ ]  Consider using [glob file lists](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#specifying-files-to-include-in-output) to split your package into separate outputs i.e. sofa, sofa-devel, libsofa so that end-users don't get headers, binaries, or /share that they don't need.

Alright. I'm gonna review my targets architecture to follow your advice.

**Comment by olivier-roussel** [comment]
>     * [ ]  REQUIRED: Non-header-only external projects including (but maybe not limited to) TinyXML, CXXOpts, Metis, should use separate conda packages with dynamic links. We already have many of these on conda-forge.

@carterbox After discussion with the developer team, would it be OK for conda-forge to have a "find package or fetch" policy for all embedded static libraries ? More precisely, cmake would first try to use an external package if available (and for the conda recipe, we would fall into that case by setting up the proper dependencies in the recipe) and then dynamic link with it. And otherwise, if the package is not foudn, cmake would fetch the package source from its repository and still static link with it (to keep compatibility with existing behavior).

**Comment by olivier-roussel** [comment]
Also, I believe CXXOpts is a header-only library, right ?
https://github.com/jarro2783/cxxopts

**Comment by olivier-roussel** [comment]
Another alternative to my previous proposition to handle currently statically linked libraries: we could consider to have a cmake option to force the use of external dependencies handled by the package manager. This might be considered is the "find of fetch" policy would not be acceptable by conda-forge as it would not enforce that the code is not fetched (and hence statically linked) if the package is not found by accident.
I guess this would be more acceptable by conda-forge, but I have to check with the development team if it is also acceptable on their side.

**Comment by xhochy** [comment]
The last approach is what most projects do and what we would like to have here

**Comment by olivier-roussel** [comment]
@conda-forge/help-c-cpp @carterbox Requested changes have been done and the recipe is ready for another round of review ! 

The recipe have been deeply changed to support multiple outputs and fix requested other points. A lot of work has been done also on SOFA code for the last release v23.12 to comply to conda-forge requirements (mainly getting out embedded libraries code to use external ones).

>    * [x]  REQUIRED: All the header-only external projects need to have their licenses added to the license section of this recipe.  https://github.com/sofa-framework/sofa/tree/master/extlibs

Fixed [here](https://github.com/conda-forge/staged-recipes/pull/23085/files#diff-e274732b2b27069fef7259193fcc1dfab914ebde0831bc632fb1d4fd046ebb74R122-R124) 

>    * [x]  REQUIRED: Non-header-only external projects including (but maybe not limited to) TinyXML, CXXOpts, Metis, should use separate conda packages with dynamic links. We already have many of these on conda-forge.

SOFA now use conda packages for these, with dynamic linking. Default behavior in SOFA for these is now to look for external packages, and if not found to fetch sources. We prevent source code fetching and restrict to use of conda package in our case (there is now a cmake option to prevent this: https://github.com/sofa-framework/sofa/blob/master/CMakeLists.txt#L39).

>    * [x]  REQUIRED: libcsparse is not licensed for redistribution? The [license file for that package ](https://github.com/sofa-framework/sofa/blob/master/Sofa/Component/LinearSolver/Direct/extlibs/csparse/COPYING.txt) is not a license. It is a copyright statement. We cannot distribute code from this project in any manner until we have a valid license.

CSparse-based linear solvers have now been moved to an external plugin (https://github.com/sofa-framework/CSparseSolvers) and is not part of the conda package, other solvers are available here (e.g. based on Eigen or Metis).

>    * [x]  Consider using [glob file lists](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#specifying-files-to-include-in-output) to split your package into separate outputs i.e. sofa, sofa-devel, libsofa so that end-users don't get headers, binaries, or /share that they don't need.

Package has been split according to your remarks and SOFA core development team, and we now provide the following output for SOFA package:
  - **libsofa-core**: SOFA core runtime libraries
  - **libsofa-core-devel**: SOFA core development files: runtime libraries + headers + cmake files (depends on libsofa-core)
  - **sofa-gl**: SOFA.GL rendering library (devel version / depends on libsofa-core)
  - **sofa-gui-qt**: Qt based GUI library for SOFA (devel version / depends on sofa-gl)
  - **sofa-app**: SOFA runtime binaries (sofaRun) + share resources (depends on sofa-gui-qt)

I did not manage to get satisfying results with file globbing method directly in the recipe. It's been a long time ago, but if my memories are correct it was rebuilding several time the same code for different outputs. So I took another approach using scripts to move files at the right place, similarly to what has been done for [boost](https://github.com/conda-forge/boost-feedstock).

External SOFA plugins packages have not been modified, as they are definitly smaller and it would result in a excessive combination. Thus this recipe still include the following conda packages for SOFA plugins, in addition the previously mentioned ones for SOFA core:
  - **sofa-python3**
  - **sofa-stlib**
  - **sofa-modelorderreduction**
  - **sofa-beamadapter**
  - **sofa-softrobots**

MacOS build is now fixed, even if we had some troubles with our sofa-python3 library which normally links with libpython as it embeds a python interpreter, which causes segv when using this sofa-python3 library within a python packaged by conda (which links statically with libpython, so we would have duplicate symbols). The problem is internal to the architecture of sofa-python3 which should split both features: python bindings and python interpreter to avoid this.

**Comment by olivier-roussel** [comment]
@wolfv Could you help reviewing this recipe ?

**Comment by carterbox** [comment]
@olivier-roussel, sorry for the wait. I think it's mostlyu just me and xhochy, and I only make the rounds about once a month.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by olivier-roussel** [comment]
The PR is waiting that multi-ouput cache feature of rattler-build to be moved out of `--experimental`, as this mode can't be enabled at staged-recipe PRs, only on existing feedstocks. If I understood correctly from the rattler core devs, this depends on https://github.com/conda/ceps/pull/102, which is still in progress.
So either we keep this PR open, either I'll re-open one once rattler build will be ready.

**Comment by olivier-roussel** [comment]
Still waiting for the previously mentioned CEP to converge. Please let this PR open or let me know if it will be more appropriated to create a new PR once multi-ouput features in rattler-build will be available.

**Comment by olivier-roussel** [comment]
Here a new pass for the recipe, hopefully making it acceptable.

I simplified the PR as it was huge and still has a long history. Now, the PR only concerns `sofa` recipe. Previous recipes for SOFA plugins (such as `sofa-python3` for python bindings & python scene reader, `sofa-app` for GUI application, etc.) will be submitted in dedicated PR once this one merged.

Another motivation to remove other recipes from here is the global build graph computation from `scripts/compute_build_graph.py` that was now returning inconsistent output in any case, making the global build impossible to solve (e.g. here: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1372083&view=logs&j=4cabe686-70ae-553a-7fd0-310379f2cbac&t=6a4fc7c9-c31a-5115-eff9-6479d72b69ff&l=676 `sofa-glfw` was required to be built first, despite it depends on `sofa-devel`). I suspect we are entering some edge cases with mixed multi-output recipes and other recipes, using rattler v1 recipe format.

As the multi-ouput cache CEP is still in progress, I rewrote the recipe as a simple multi-output recipe, without cache so no experimental features from rattler-build. Drawbacks are:
- duplicate recipe dependencies for `libsofa` and `sofa-devel`
- duplicate build for `libsofa` and `sofa-devel`

We should be able to solve that when cache feature will be available.

To summarize, this PR for the unique `sofa` multi-output recipe has now as outputs (following our previous discussions about splitting `sofa`):
- **libsofa**: runtime libraries for SOFA core
- **sofa-devel**: libsofa + development headers & cmake configuration files
- **sofa-gl**: SOFA GL rendering library (which is part of SOFA same source tarball). It is also a common library for other SOFA rendering plugins such as `sofa-glfw` GUI or `sofa-qt`, which has be removed from this PR to be submitted later.  Eventually, this could be moved out as a separate recipe, if you prefer this way.

Also, version of SOFA has been bumped to latest release v25.06. 

@carterbox @xhochy This is ready for a new review. Many thanks !

File: recipes/sofa/recipe.yaml [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+          then: ninja
+      host:
+        - libboost-headers
+        - zlib
+        - tinyxml2
+        - eigen
+        - cxxopts
+        - gtest
+      run:
+        - zlib
```
Comment (traversaro) [inline recipes/sofa/recipe.yaml]: If this is required for `find_package(SOFA)` to succeed, perhaps it should be in `sofa-devel` ?

File: recipes/sofa/recipe.yaml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    about:
+      summary: Runtime librairies for the SOFA framework
+      description: Runtime librairies for SOFA, a real-time multi-physics
+       simulator with an emphasis on medical simulation and robotics.
+      license: LGPL-2.1-or-later
+      license_file:
+        - LICENSE-LGPL.md
+        - extlibs/difflib/LICENSE.MIT
+        - extlibs/json/LICENSE.MIT
+        - extlibs/stb/LICENSE.MIT
```
Comment (traversaro) [inline recipes/sofa/recipe.yaml:70]: I think this is fine, but can you explain why can't this be unvendored? Except for `stb` that I already know why.

File: recipes/sofa/recipe.yaml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        - ${{ pin_subpackage('libsofa', exact=True) }}
+      run_exports:
+        - ${{ pin_subpackage('libsofa', upper_bound='x.x.x') }}
+    tests:
+      - package_contents:
+          include:
+            - Sofa.Core/sofa/core/init.h
+          files:
+            - ${{ "Library/" if win }}lib/cmake/Sofa.Core/Sofa.CoreConfig.cmake
+            - ${{ "Library/" if win }}lib/cmake/Sofa.Core/Sofa.CoreConfigVersion.cmake
```
Comment (traversaro) [inline recipes/sofa/recipe.yaml]: Not necessary, but if you want to ensure that the `find_package(Sofa.Core)` or similar works fine and check if some targets exists, you can also use https://github.com/ami-iit/cmake-package-check .

File: recipes/sofa/recipe.yaml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+      build:
+        - ${{ compiler('cxx') }}
+        - ${{ stdlib('c') }}
+        - cmake
+        - if: unix
+          then: make
+        - if: win 
+          then: ninja
+        - if: linux
+          then: ${{ cdt('mesa-libgl-devel') }}
```
Comment (traversaro) [inline recipes/sofa/recipe.yaml]: The use of CDT for GL packages is now superseded, you can just depend on `libgl-devel` as a normal host package, and everything should work fine.

Comment (olivier-roussel) [reply recipes/sofa/recipe.yaml]: Oh I've misunderstood that, thanks, that is now clear. Fixed.

File: recipes/sofa/recipe.yaml (hunk 3 above)
Comment (olivier-roussel) [reply recipes/sofa/recipe.yaml]: Thanks for the tool, that was actually missing and such cmake minimalist test programs were needed to be written manually (which was not done here). I've included it and thanks to it could easily fix missing run dependencies.

File: recipes/sofa/recipe.yaml (hunk 2 above)
Comment (olivier-roussel) [reply recipes/sofa/recipe.yaml:70]: Initially, there was many vendored software in SOFA. I've pushed the devs to use external dependencies as much as possible, but some cases such as header-only were not priorized. 
And among these three, `json` (`nlohmann_json` to be precise here), can indeed be unvendored. I've added a patch to the recipe for this and the corresponding PR for the next SOFA release can be found here https://github.com/sofa-framework/sofa/pull/5775.
For `difflib`, it might not be clear but here we are dealing with https://github.com/duckie/difflib (and not the python package), which I believe to not be available on conda-forge and as it is unmaintened for 11 years, we probably do not want to add it neither.

File: recipes/sofa/recipe.yaml (hunk 1 above)
Comment (olivier-roussel) [reply recipes/sofa/recipe.yaml]: You are right. It should not be there, and it should not be necessary in `sofa-devel` if I understood correctly as it defines some `run-exports`.

**Comment by olivier-roussel** [comment]
Many thanks @traversaro for your review ! All requested changes should have been made now, this is ready for another round or merging !

**Comment by traversaro** [comment]
Thanks @olivier-roussel ! Can I rebase and squash, or you prefer to curate the commit yourself?

**Comment by olivier-roussel** [comment]
@traversaro Sure feel free to rebase & squash as you wish. Thanks again for your help, this PR was lasting for too long (more than 2 years) and involved so much energy.

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
sorry, I see that there is a still an unresolved review from @carterbox , can you find his comments so we can check if we address them? I can't find them at the moment.

**Comment by olivier-roussel** [comment]
Indeed the github interface is not very helpful to find out where exactly are the requested changes mentioned in the "change requested" summary. To my understanding, @carterbox requested changes twice, and this should concern his second review ( https://github.com/conda-forge/staged-recipes/pull/23085#pullrequestreview-2039131894 ). 
To me, everything is addressed for a while now, the recipe has quite changed since.
Is there a way we can still progress on this PR despite he seems unavailable (I requested a new review exactly one month ago) ?

**Comment by traversaro** [comment]
> Indeed the github interface is not very helpful to find out where exactly are the requested changes mentioned in the "change requested" summary. To my understanding, @carterbox requested changes twice, and this should concern his second review ( [#23085 (review)](https://github.com/conda-forge/staged-recipes/pull/23085#pullrequestreview-2039131894) ). To me, everything is addressed for a while now, the recipe has quite changed since. Is there a way we can still progress on this PR despite he seems unavailable (I requested a new review exactly one month ago) ?

Thanks for the link. This review indeed seems solved, by creating the `sofa-devel` package and by installing directly in `$CONDA_PREFIX`. I would wait for @carterbox confirmation on this for a few days, and if he do not reply, feel free to ping me again so I can merge. Like 2/3 working days I think it is reasonable, does that sound ok?

**Comment by olivier-roussel** [comment]
That's great, thanks again !

**Comment by olivier-roussel** [comment]
Hi @traversaro, as we had no feedback could this be merged ?

**Comment by traversaro** [comment]
Ok!
